kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/TopicNameRetriever.java

```diff
@@ -32,23 +33,31 @@
 import io.kroxylicious.proxy.tag.VisibleForTesting;
 
 import static java.util.Collections.unmodifiableMap;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.concurrent.CompletableFuture.supplyAsync;
 import static java.util.stream.Collectors.toMap;
 
 final class TopicNameRetriever {
     // Version 12 was the first version that uses topic ids.
     private static final short METADATA_API_VER_WITH_TOPIC_ID_SUPPORT = (short) 12;
     private final FilterContext filterContext;
     private final Executor filterDispatchExecutor;
+    private final BooleanSupplier inFilterDispatchThread;
 
-    TopicNameRetriever(FilterContext filterContext, Executor filterDispatchExecutor) {
+    TopicNameRetriever(FilterContext filterContext, Executor filterDispatchExecutor, BooleanSupplier inFilterDispatchThread) {
         this.filterContext = filterContext;
         this.filterDispatchExecutor = filterDispatchExecutor;
+        this.inFilterDispatchThread = inFilterDispatchThread;
     }
 
     CompletionStage<TopicNameMapping> topicNames(Collection<Uuid> topicIds) {
         Objects.requireNonNull(topicIds);
         if (topicIds.isEmpty()) {
-            return CompletableFuture.supplyAsync(() -> MapTopicNameMapping.EMPTY, filterDispatchExecutor).minimalCompletionStage();
+            // if we are already in the filter dispatch thread, we can return the completed future directly
+            CompletableFuture<TopicNameMapping> responseStage = inFilterDispatchThread.getAsBoolean()
```

Comment:
Does it need to be this complicated? AFAICS topicNames is called directly after an retriever instance is created, so we can just pass a boolean, rather than a `BooleanSupplier` can't we?